Testnet/NFTAuctionStore/NFTAuctionStore/NFTAuctionStore.cs

```diff
@@ -0,0 +1,199 @@
+        : base(state)
+    {
+        EnsureNotPayable();
+    }
+
+    public void Auction(Address contract, ulong tokenId, ulong startingPrice, ulong duration)
+    {
+        EnsureNotPayable();
+
+        var tokenOwner = GetOwner(contract, tokenId);
+
+        Assert(tokenOwner != Address, "The token is already on sale.");
+
+        EnsureCallerCanOperate(contract, tokenOwner);
+
+        TransferToken(contract, tokenId, tokenOwner, Address);
+
+        var auction = new AuctionInfo
+        {
+            Seller = tokenOwner,
+            EndBlock = checked(Block.Number + duration),
+            StartingPrice = startingPrice,
+            Ended = false
+        };
+
+        SetAuctionInfo(contract, tokenId, auction);
+        Log(new AuctionStartedLog { Contract = contract, TokenId = tokenId, EndBlock = auction.EndBlock, Seller = auction.Seller, startingPrice = auction.StartingPrice });
+    }
+
+    public void Bid(Address contract, ulong tokenId)
+    {
+        var auction = GetAuctionInfo(contract, tokenId);
+
+        Assert(Block.Number < auction.EndBlock, "Auction ended.");
+
+        Assert(Message.Value > auction.HighestBid && Message.Value >= auction.StartingPrice, "The amount is not higher than highest bidder or starting price.");
+
+        if (auction.HighestBid > 0)
+        {
+            //refund for previous bidder
+            var balance = GetRefund(auction.HighestBidder);
+
+            SetRefund(auction.HighestBidder, balance + auction.HighestBid);
+        }
+
+        auction.HighestBidder = Message.Sender;
+        auction.HighestBid = Message.Value;
+
+        SetAuctionInfo(contract, tokenId, auction);
+
+        Log(new HighestBidUpdatedLog { Contract = contract, TokenId = tokenId, Bidder = auction.HighestBidder, Bid = auction.HighestBid });
+    }
+
+    public bool Refund()
+    {
+        EnsureNotPayable();
+
+        var amount = GetRefund(Message.Sender);
+
+        Assert(amount > 0);
+
+        SetRefund(Message.Sender, 0);
+
+        var transfer = Transfer(this.Message.Sender, amount);
+
+        if (!transfer.Success)
+            SetRefund(Message.Sender, amount);
+
+        return transfer.Success;
+    }
+
+    public void AuctionEnd(Address contract, ulong tokenId)
+    {
+        EnsureNotPayable();
+
+        var auction = GetAuctionInfo(contract, tokenId);
+
+        Assert(Block.Number >= auction.EndBlock, "Auction is not ended yet.");
+
+        Assert(!auction.Ended,"Auction end already executed.");
+
+        auction.Ended = true;
+        SetAuctionInfo(contract, tokenId, auction);
+
+        if (auction.HighestBid > 0)
+        {
+            var result = Transfer(auction.Seller, auction.HighestBid);
+
+            Assert(result.Success, "Transfer failed.");
+
+            SafeTransferToken(contract, tokenId, Address, auction.HighestBidder);
+        }
+        else
+        {
+            SafeTransferToken(contract, tokenId, Address, auction.Seller);
+        }
```

Comment:
Not entirely accurate, if the starting price was higher than 0, the seller would never get their token back.